Added size comparison

Tiberriver256 opened the pull request:
When MatchSource is $False and Test-Path is $True added a webclient
content-length check to compare the size in bytes of the target and
destination files.

This will ensure that the file is re-downloaded if the file sizes are
not the same (i.e. target file has been updated, download was
interrupted previously, local file was corrupted, etc.)

msftclas commented:
Hi __@Tiberriver256__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, MSBOT;

Tiberriver256 commented:
Looks much better now! :)

KarolKaczmarek commented on DSCResources/MSFT_xRemoteFile/MSFT_xRemoteFile.psm1:
```diff
@@ -252,8 +252,15 @@ function Test-TargetResource
             } else {
-                Write-Debug "MatchSource is false. No need for downloading file."
-                $fileExists = $true 
+                Write-Debug "Destination file size is not the same as source file. File will be downloaded."
```
Not quite, you need to put it in the else block under line 263:
```
if($bytes_total -eq (Get-Item $DestinationPath).Length)  
            {  
                   Write-Debug "MatchSource is false. No need for downloading file."  
                   $fileExists = $true  
             }  
else
{
Write-Debug "Destination file size is not the same as source file. File will be downloaded." 
}

KarolKaczmarek commented on DSCResources/MSFT_xRemoteFile/MSFT_xRemoteFile.psm1:
```diff
@@ -252,8 +252,15 @@ function Test-TargetResource
+                [Long]$bytes_total = $wc.ResponseHeaders["Content-Length"]
+                if($bytes_total -eq (Get-Item $DestinationPath).Length)
+                {
+                    Write-Debug "MatchSource is false. No need for downloading file."
```
I'd also modify this message to make it more accurate, something like:
"Size of the destination and source files match. No need for downloading the file"

KarolKaczmarek commented:
Thanks, I added some comments to your changes, however looking at the code again, I think we should take a different approach.
Currently all -matchsource $true does is checking cache which really just stores the last write time. We should simply add size of the file as another entry of the cache (see Set-TargetResource function, line 190) to perform verification of yet another property of the file when using -matchsource (we can't literally compare the source - content - in this case since that would require actually downloading the entire file to decide whether it should be downloaded...).

Does it make sense?

Tiberriver256 commented:
I added your recommended changes to the second write-debug statement and moved the first write-debug to the proper spot.

I'm not really sure what you mean when you say you cannot compare the file size of the source and the destination file. That is what the code there does without downloading the entire file. You can always see how large a file is before you start downloading.

See here: http://www.c-sharpcorner.com/blogs/get-file-size-before-downloading-using-webclient1

KarolKaczmarek commented:
That's not what I meant. I said you can't compare the source (content) of the file, not the size.

Take a look at these lines:
```
# Update cache 
190 190      if (Test-Path $DestinationPath) 
191 191      { 
192 192          $downloadedFile = Get-Item $DestinationPath 
193 193          $lastWriteTime = $downloadedFile.LastWriteTimeUtc 
194 194          $inputObject = @{} 
195 195          $inputObject["LastWriteTime"] = $lastWriteTime 
196 196          Update-Cache -DestinationPath $DestinationPath -Uri $Uri -InputObject $inputObject 
197 197      }    
```

All what I'm saying is that we should be storing the size as another entry in the cache:
```
$inputObject["FileSize"] = $fileSize
```
 and later compare it in the test-targetresource method similariily how we do it with the last write time currently.

Tiberriver256 commented:
Oh, okay. I got you now. That would make for much cleaner code, yes.

I will make the appropriate changes and re-commit. Thank you for the help looking this over.

KarolKaczmarek commented:
Great.
So just to make sure we are on the same page, my proposition is to come back to the original behavior for matchsource=$false (and if that's the case, we simply do:
```powershell
Write-Debug "MatchSource is false. No need for downloading file."  
$fileExists = $true   
```

and for matchsource=$true, we should add the extra check for the size here:

```powershell
if ($MatchSource) {            
241 241                  $file = Get-Item $DestinationPath 
242 242                  # Getting cache. It's cleared every time user runs Start-DscConfiguration 
243 243                  $cache = Get-Cache -DestinationPath $DestinationPath -Uri $Uri 
244 244   
245 245                  if ($cache -ne $null -and ($cache.LastWriteTime -eq $file.LastWriteTimeUtc)) 
246 246                  { 
247 247                      Write-Debug "Cache reflects current state. No need for downloading file." 
248 248                      $fileExists = $true 
249 249                  } 
# TODO we should be checking if the file size changed
```

Tiberriver256 commented:
Would this make sense?

```
        $downloadedFile = Get-Item $DestinationPath
        $lastWriteTime = $downloadedFile.LastWriteTimeUtc
        $RemoteFileRequest = [System.Net.WebClient]::New().OpenRead($URI)
        $filesize = $RemoteFileRequest.ResponseHeaders["Content-Length"]
        $inputObject = @{}
        $inputObject["LastWriteTime"] = $lastWriteTime
        $inputObject["FileSize"] = $filesize
        Update-Cache -DestinationPath $DestinationPath -Uri $Uri -InputObject $inputObject
```

and

```
if ($cache -ne $null -and ($cache.LastWriteTime -eq $file.LastWriteTimeUtc) -and ($cache.FileSize -eq $file.Length))
                {
                    Write-Debug "Cache reflects current state. No need for downloading file."
                    $fileExists = $true
                }
```

Tiberriver256 commented:
Ended up going with a simpler approach and instead of getting the file size from the source and just updating the cache using the size of the downloaded file.

How is it looking now? Thanks for working on this with me by the way. I'm still fairly new to coding in general so this helps me learn about clean coding and better long-term thinking.

KarolKaczmarek commented:
Looks good now, thanks!
